### ISSUE TITLE
Tuning, stage 2, of candidate results - projecting to summary DTO's instead of entities  

### ISSUE THREAD

**Issue by camerojo** [talentbeyondboundaries/tbbtalentv2#1286]
Performance of loading candidate information in response an API request for a set of candidates is dependant on the number of database accesses required to fetch the required data.

This issue is based on issue #1284 (ie it contains all the changes there). Issue #1284 reduced the number of accesses mainly by not fetching associated candidate related data if it was not required in the front end display of the summary candidate display (ie displaying the candidate in a single line in the list of results). This is because each type of candidate related table data (ie data store in a table other than the candidate table) typically requires one or more additional database accesses to fetch it.

This issue seeks performance improvements by not loading candidate entities from the database, but instead loading a special CandidateSummary object (DTO - Data Transfer Object).

This has the potential for further performance improvement by:
- gathering all summary data in a single database access - by computing (in the database query) the potentially displayed data that is dependent on associated tables without actually needing to load those tables (Instead of joins fetch calculated fields from other tables in subselect - eg select id, (select x from other-table o where o.candidate_id=id) as xfield from candidate.
- loading into non entity DTOs - eg CandidateSummary. Loading data into entity objects when you only need the data for readonly purposes is unnecessary and inefficient. See https://thorben-janssen.com/entities-dtos-use-projection/ 

This will make use of the the new SearchCandidateRequest.extractSQL method which identifies the candidates to be returned.

Note that this DTO approach allows us to directly return fully documented Java classes from our API, rather than Maps and Lists of Maps. For example we can return List\<CandidateSummary\>. See #97

Note also that the filtering of data, based on what the logged in user can see can now be applied BEFORE the database fetching instead of after in the DtoBuilder. This is more logical and efficient since we do not fetch data we are not going to use.

**Comment by camerojo** [comment]
### Trial the summary approach
Trial using the example of tasks. Candidate summary display can display following summary task information:
- total number of tasks
- number of completed tasks
- number of overdue tasks
- number of abandoned tasks

- [ ] Compute all above values in database query
- [ ] Build CandidateSummary DTO which contains candidate summary data including the above computed data
- [ ] Fetch that CandidateSummary DTO in a single database access, constructing the query in SQL based on the fetch SQL returned by the new SearchCandidateRequest.extractSQL method.

**Comment by cazcam34** [comment]
When in candidate portal, I am following the existing code pattern and calling the CandidatePortalApi to getCandidateLoadXXX. Eg. getCandidateLoadDestinations. This calls a custom query to fetch the lazy destinations and join on the candidate object. What I noticed is that it seems that the all lazy loaded objects are on the candidate object when calling this method, and I can even get them just by calling findById. I found the article below which led me to think perhaps we aren't lazy loading child entities? Food for thought...
https://medium.com/javarevisited/spring-data-jpa-findbyid-anti-pattern-aefbf045fc44

**Comment by camerojo** [comment]
CandidateDtoService as distinct from the Entity focussed CandidateService

See the QueryHelper 

Could get all candidates in one go
Then all 1-n sub entities of a given class in one go
Then put it all together

So the number of db accesses is the number of 1-n entities + 1 (for the candidate itself)

Another approach which allows for single returned row per candidate 

https://chatgpt.com/s/t_6945c3e4b9288191b07e2f841202479a

### API and DtoBuilder

Defer returning dto in API because we still really need the DtoBuilder for filtering. Fetch dtos (not entities) from DB using CandidateDtoService then still pass those objects through the DtoBuilder for filtering - so API still returns List of Maps.

**Comment by camerojo** [comment]
- [x] Need UserDto
- [x] Need PartnerDto
- [x] Look at api dtos. Can I use them?
- [x] Processing the dto to build the sql query? See https://chatgpt.com/s/t_69471e9472f08191a4e17b7e0a1c43f9
- [x] Add fast search checkbox (ignores old search checkbox)
- [x] Annotate dtos and complete unit tests of sql generation.
- [x] Error ERROR: cannot pass more than 100 arguments to a function - can chunk up multiple calls to jsonb_build_object
- [x] With Accountants search Getting error Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.lang.Boolean` from String "ValidPassport": only "true" or "false" recognized
- [x] No error with Admin assitanct.
- [x] Return Map id, CandidateReadDto from DAO makes it easy to reorder according to originally sorted ids.
- [x] Implement cached table on server
- [x] Convert new cache code to use LogBuilder
- [x] Get markUserSelectedCandidates working - see todo in searchCandidateDtos of SavedSearchService
- [x] Can I map Strings to DTO enum types in json building
- [x] Test candidate version updates
- [x] url is Computed field in Attachment entity and dto - see todo in CandidateAttachmentReadDto
- [x] Does Redis memory caching make a difference? See todo in CandidateRedisCache - there is no TTL.
- [ ] Not showing job experience descriptions or candidate text - in dtos - maybe because there is no occupation. Instead of deleting Circular reference, add mini dto for Candidate Occupation in Candidate Job Reference. Existing Angular code (dtotype != api) has cj referencing co. Why? Api logic makes more sense. See issue #2950
- [ ] Circular dependence in db - CandidateJobExperience <--> CandidateOccupation. Had to remove from dtos. See todo in CandidateJobExperienceReadDto.  See #2950


Remaining issues moved to issue #2968 

- [ ] Why do we need SalesforceJobOppReadDto (used in CandidateVisaJobCheck) and JobOppReadDto
- [ ] numberDependants, pendingTerms, shareableCv, shareableDoc, contextNote are computed fields in Candidate - see todo in CandidateReadDto
- [ ] Missing taskAssignments see todo in CandidateReadDto
- [ ] Doesn't seem to handle two 1-many: candidateVisaChecks - candidateVisaJobChecks. Test new SqlJsonQueryBuilder. Remove Todo commenting out candidateVisaChecks in CandidateReadDto
- [ ] CandidateVisaJobCheckReadDto - compare with candidateVisaJobCheckDto in CandidateBuilderSelector.java. Check Angular whether SalesforceJobOppReadDto is or JobOppReadDto is needed   (see also  candidateVisaJobDto in CandidateVisaJobCheckAdminApi.java)
- [ ] Convert all DTOs to use enum types instead of Strings when needed
- [ ] Convert all AttachmentType.file locations in CandidateAttachment to urls. See getUrl code in the CandidateAttachment entity. Then can update the CandidateAttachment entity to just return location when url is requested. 



###To do
TL;DR (the mental checklist)

Before trusting the cache:

✅ Every table that affects candidate JSON bumps candidate.data_version

✅ Indirect tables use JOIN-based triggers

✅ coalesce(NEW, OLD) everywhere

✅ Users trigger candidate bumps

✅ AFTER triggers only

✅ Candidate deletes clean up cache

**Comment by camerojo** [comment]
### Summary of issues and planned approach

- Note that we could not have done this earlier while we were still using Criteria Specifications. We needed to move to the pure SQL approach that we introduced in 2.4.0
- Retrieving Candidate's is the main issue. Candidates have a lot of joined tables which raises special challenges. That design is OK - it is the natural design for a relational data base. But some issues arise with entities which have a lot of joined tables.
- Entities are not ideally suited for DTOs. The power of entities is around their ability to automatically turn Java entity updates into SQL requests. But for read only use - eg DTOs - entities serve little purpose and add unnecessary complexity, constraints and performance overhead. Simple entities are OK to be used as DTO's - but complex entities such as the Candidate entity should be mapped to traditional DTOs.
- Entities normally "lazy load" joined attributes. But this is not desirable when generating search results. It leads to multiple unnecessary db accesses which impacts performance.
- Temporarily overriding lazy loading for searches is possible but that leads to multi bag errors and poor performance due to Cartesian products.
- Spring projections can create DTOs from a JPA repository but with multiple joins you can run in to multiple bag errors and Cartesian product performance issues.
- Normal db access methods return multiple rows per candidate. 

The planned approach is to avoid all the above issues by using Postgres's jsondb support to encode each joined table attribute into a single String field.

This means that each candidate is returned as a single row. A whole page of candidates can be returned by a single db access.

Then the encoded json strings can be decoded into structured DTO objects (or Lists of DTOs in the case of 1 to many relationships) on the server.

This should greatly improve performance, particularly for candidate search.

For now we can continue using DtoBuilders to process the returned Candidate DTO objects, mapping those objects onto our existing List of Maps API, while applying the filtering logic that is encapsulated in the DtoBuilders.

However, a future improvement could directly return Candidate DTO objects in the API. It is not suggested that we tackle that yet. This issue is about improving performance - not improving our internal API.
